From ebd841d0973df76eb19eaf01452645ca190c4671 Mon Sep 17 00:00:00 2001
From: Venkataramana Kotakonda <venkataramana.kotakonda@intel.com>
Date: Thu, 24 Dec 2015 13:27:14 +0530
Subject: [PATCH 06/14] USBC/DPM: Handle usb typec phy notifications in dpm.

DPM adds notification handler to phy to received phy notifications
like vbus change and pass it to policy engine.

Change-Id: I460117c9f41ba2f4feb0f3dc6a7a37df77cf1c3c
Tracked-On: https://jira01.devtools.intel.com/browse/OAM-15496
Signed-off-by: Venkataramana Kotakonda <venkataramana.kotakonda@intel.com>
Reviewed-on: https://android.intel.com:443/452121
---
 drivers/usb/typec/pd/devpolicy_mgr.c | 25 +++++++++++++++++++++++++
 drivers/usb/typec/pd/devpolicy_mgr.h |  5 +++++
 2 files changed, 30 insertions(+)

diff --git a/drivers/usb/typec/pd/devpolicy_mgr.c b/drivers/usb/typec/pd/devpolicy_mgr.c
index 268dcf0..e8d4f72 100644
--- a/drivers/usb/typec/pd/devpolicy_mgr.c
+++ b/drivers/usb/typec/pd/devpolicy_mgr.c
@@ -574,6 +574,31 @@ static int dpm_set_display_port_state(struct devpolicy_mgr *dpm,
 	return 0;
 }
 
+void dpm_handle_phy_event(struct typec_phy *phy,
+				enum typec_phy_dpm_evts evt)
+{
+	enum devpolicy_mgr_events dpm_evt = DEVMGR_EVENT_NONE;
+
+	if (!phy || !phy->proto || !phy->proto->p || !phy->proto->p->dpm)
+		return;
+
+	switch (evt) {
+	case PHY_DPM_EVENT_VBUS_ON:
+		dpm_evt = DEVMGR_EVENT_VBUS_ON;
+		break;
+	case PHY_DPM_EVENT_VBUS_OFF:
+		dpm_evt = DEVMGR_EVENT_VBUS_OFF;
+		break;
+	default:
+		pr_info("DPM:%s: Unknown phy event=%d", __func__, evt);
+	}
+
+	if (dpm_evt != DEVMGR_EVENT_NONE)
+		dpm_notify_policy_evt(phy->proto->p->dpm, dpm_evt);
+
+}
+EXPORT_SYMBOL(dpm_handle_phy_event);
+
 static void dpm_handle_ext_cable_event(struct devpolicy_mgr *dpm,
 					struct cable_event *evt)
 {
diff --git a/drivers/usb/typec/pd/devpolicy_mgr.h b/drivers/usb/typec/pd/devpolicy_mgr.h
index 7d72ec0..2026c1a 100644
--- a/drivers/usb/typec/pd/devpolicy_mgr.h
+++ b/drivers/usb/typec/pd/devpolicy_mgr.h
@@ -410,6 +410,8 @@ extern struct class *power_delivery_class;
 extern struct devpolicy_mgr *dpm_register_syspolicy(struct typec_phy *phy,
 				struct pd_policy *policy);
 extern void dpm_unregister_syspolicy(struct devpolicy_mgr *dpm);
+extern void dpm_handle_phy_event(struct typec_phy *phy,
+					enum typec_phy_dpm_evts evt);
 #else /* CONFIG_USBC_PD && CONFIG_USBC_PD_POLICY */
 static inline
 struct devpolicy_mgr *dpm_register_syspolicy(struct typec_phy *phy,
@@ -419,6 +421,9 @@ struct devpolicy_mgr *dpm_register_syspolicy(struct typec_phy *phy,
 }
 static inline void dpm_unregister_syspolicy(struct devpolicy_mgr *dpm)
 { }
+static inline void dpm_handle_phy_event(struct typec_phy *phy,
+					enum typec_phy_dpm_evts evt)
+{ }
 #endif /* CONFIG_USBC_PD && CONFIG_USBC_PD_POLICY */
 
 #endif /* __PD_DEVMGR_POLICY_H__ */
-- 
1.9.1

